src/charm.py

```diff
@@ -0,0 +1,100 @@
+
+    def __init__(self, framework: ops.Framework):
+        super().__init__(framework)
+        self.container = self.unit.get_container("gocert")
+        self.tls = TLSCertificatesProvidesV3(self, relationship_name="certificates")
+
+        framework.observe(self.on["gocert"].pebble_ready, self._on_gocert_pebble_ready)
+        framework.observe(self.on["gocert"].pebble_custom_notice, self._on_gocert_notify)
+        framework.observe(self.tls.on.certificate_creation_request, self._on_new_certificate)
+        framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.on.collect_unit_status, self._on_collect_status)
+
+    def _on_gocert_pebble_ready(self, event: ops.PebbleReadyEvent):
+        try:
+            self.container.pull("/etc/config/config.yaml")
+        except ops.pebble.PathError:
+            config_file = open("src/configs/config.yaml").read()
+            cert_file = open("src/configs/cert.pem").read()
+            key_file = open("src/configs/key.pem").read()
+            self.container.make_dir(path="/etc/config", make_parents=True)
+            self.container.push(path="/etc/config/config.yaml", source=config_file)
+            self.container.push(path="/etc/config/cert.pem", source=cert_file)
+            self.container.push(path="/etc/config/key.pem", source=key_file)
+            self.container.add_layer("gocert", self._pebble_layer, combine=True)
+            self.container.replan()
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent):
+        event.add_status(ops.ActiveStatus())
+
+    def _on_new_certificate(self, event: CertificateCreationRequestEvent):
+        csr = event.certificate_signing_request
+        requests.post(
+            url="https://localhost:2111/api/v1/certificate_requests",
+            data=csr,
+            headers={"Content-Type": "text/plain"},
+            verify=False,
+        )
+
+    def _on_gocert_notify(self, event: ops.PebbleCustomNoticeEvent):
+        r = requests.get(url="https://localhost:2111/api/v1/certificate_requests", verify=False)
+        response_data = r.json()
+        for relation in self.model.relations["certificates"]:
+            unfulfilled_csrs = self.tls.get_outstanding_certificate_requests(
+                relation_id=relation.id
+            )
+            for row in response_data:
+                csr = row.get("CSR")
+                cert = row.get("Certificate")
+                for relation_csr in unfulfilled_csrs:
+                    if csr == relation_csr.csr:
+                        self.tls.set_relation_certificate(
+                            certificate_signing_request=csr,
+                            certificate=cert,
+                            ca="",
+                            chain=[cert],
+                            relation_id=relation.id,
+                        )
+
+    def _on_config_changed(self, event: ops.ConfigChangedEvent):
+        self.unit.status = ops.ActiveStatus()
```

Comment:
Status should be set on collect unit status only, I don't think we need this event handler at all, we can redirect both `pebble ready` and `config changed` events to an `_configure` handler

Reply:
done

Reply:
just found out also, that if we don't do this `harness` can not see the unit status if we do `event.add_status(ops.ActiveStatus())`, only when we do `self.unit.status = ops.ActiveStatus()`

I moved the tests over to [scenario](https://pypi.org/project/ops-scenario/), which seems to be where charm testing is headed.